added lock feature

danger-ahead opened the pull request:
- [x] tested on Android
- [ ] tested on iOS
- [x] added all the necessary code changes for both Android and iOS for the feature to work

Lets the user to use biometrics as well as pin or pattern.
Also adds the ability to lock or unlock the app from the Profile section.

Uses `shared_preferences` to check if the user has turned on locking.
Uses `local_auth` for the actual authorisation feature to work.

Addresses #2

danger-ahead committed:
added biometric/pin/pattern lock feature

danger-ahead commented:
Do not merge this now. It has a bug. Will fix this tomorrow.

31Carlton7 commented:
So far so good! However, is it possible that you could use [hive](https://pub.dev/packages/hive) instead of shared_preferences for local storage, as hive's performance is much better and I just recently migrated the app from using shared_preferences to hive.

danger-ahead commented:
> is it possible that you could use [hive](https://pub.dev/packages/hive) 

Okay will be changing that too.

danger-ahead committed:
ADDED CONDITIONALS WHILE AUTHENTICATING

Changed Android app package name to `com.notes_app`
Changed iOS Face ID reason to a proper reason

danger-ahead commented:
@31Carlton7 Don't merge this PR yet. This needs further testing and polishing.

danger-ahead commented:
> @31Carlton7 Don't merge this PR yet. This needs further testing and polishing.

Hi @31Carlton7 
Would you please test this PR on your emulator (or device)?

Actually I think it's much consistent to work with `local_auth` if we're working with a stateful widget. However, in our case, the root widget of our app is a stateless widget.
Though I think it's working fine. Let me know if you encounter any error or bug.
Also one possible fix could be migrating the root widget to a stateful widget (I might be wrong here, just learning Flutter).

danger-ahead committed:
CHANGED THE PACKAGE NAME IN DEBUG AND PROFILE BUILDS [ANDROID]

danger-ahead committed:
REMOVED DEPRECATED API & FIXED CHANGE-LOCK-STATUS BUG

- Replaced a deprecated API in AndroidManifest.xml
- Fixed a bug in `authorisation.dart` that prevented the user from changing
  lock status

danger-ahead committed:
RESTRUCTURED AUTHENTICATION WINDOW LOADING
- Moved authenticate() under home_view.dart

danger-ahead commented:
@31Carlton7 I tried to fix most of the bugs.

One still remains:
<img src="https://user-images.githubusercontent.com/55531939/142023959-0a21f515-61f9-413a-86a6-42c37e2683b7.png" width="200">

Here, the contents of the top most note remains visible prior to authorization. I'm not much well accustomed to state management. Perhaps this bug can be solved by delaying the render of `home_view` till `authenticate()` has returned `true`. This portion resides in `home_view.dart`.

Also you can now MERGE this PR.

danger-ahead committed:
REMOVED stickyAuth
- This is a bug noticed in Android.
For more reference: https://github.com/flutter/flutter/issues/83773

31Carlton7 commented:
I'll take a look at it either tonight or tomorrow!

danger-ahead commented:
> I'll take a look at it either tonight or tomorrow!

I didn't change back to `hive` because only one pair of key-value is being used for the locking algorithm. And `shared_preferences` performs better when it comes to reading data. I found out about this from [here](https://medium.com/flutter-community/using-hive-instead-of-sharedpreferences-for-storing-preferences-2d98c9db930f).

31Carlton7 commented:
I just went through the code and tested it, and the feature doesn't work. None of the notes are locked, and it didn't exactly do anything. Please review your code and update the app to lock the notes and require password auth when trying to view a note.

danger-ahead commented:
https://user-images.githubusercontent.com/55531939/142241217-6ae115ca-8431-4a4e-aec6-602af35ea494.mp4

I tested this on multiple devices.
Which device are you using? And what's the error log?
Kindly upload the console log.

danger-ahead commented:
@31Carlton7 awaiting your reply.
